### PR TITLE
IdModel: Add the Permissive graph

### PR DESCRIPTION
**Pull request by naoyam**
Added the Permissive graph. The main logic is at `IdModel::buildPermissiveMap`. 

While validating the generated graphs by comparing them with the corresponding ComputeAtMap Permissive maps, some mismatches were found. It turned out that `ValGraph::exprsMap` needed more thorough comparisons when back propagating Merge expressions. See `ValGraph::mapMergeBackward`.

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

File: csrc/id_model/id_model.h [hunk 1]
```diff
@@ -46,6 +46,17 @@ class ValGraph;
...
+//   Forward broadcast axes in replay
+//   Map all iteration domains
+//   Always contain root mappings (otherwise they could have been forwarded in
+//   broadcast)
+// IdMappingMode::AlmostExact
+//   Forward through broadcast axes, but not through to a non-broadcast axis
+//     i.e. id{b1*i0}, id{i0} are mapped
+//          id{i1*i0}, id{i0} are not mapped (this part is the difference from
+//          PERMISSIVE)
+//   Forward through split one axes, i.e. id{ceilDiv(i0, 1)}, id{i0} are mapped
```
Comment (zasdfgbnm) [inline csrc/id_model/id_model.h:59]: Do we also forward through merge-one axes?

File: csrc/val_graph.cpp [hunk 2]
```diff
@@ -256,27 +256,8 @@ bool ValGraph::exprsMap(Expr* first, Expr* second, bool forward) const {
...
-         extent_0o->evaluate() == extent_1o->evaluate()) ||
-        disjointValSets().permissiveAreMapped(merge0->outer(), merge1->outer());
-
-    auto extent_i_match = extent_0i->sameAs(extent_1i) ||
-        (extent_0i->isConstInt() && extent_1i->isConstInt() &&
-         extent_0i->evaluate() == extent_1i->evaluate()) ||
-        disjointValSets().permissiveAreMapped(merge0->inner(), merge1->inner());
-
-    if (!(extent_o_match || extent_i_match)) {
+    if (!ValGraph::mapMergeBackward<Val>(
```
Comment (zasdfgbnm) [inline csrc/val_graph.cpp:259]: nit:
```suggestion
    if (!mapMergeBackward<Val>(
```

File: csrc/val_graph.h [hunk 3]
```diff
@@ -159,6 +159,82 @@ class ValGraph {
   // be the only call in ValGraph to mapThroughExpr.
   void maybeMapThroughExprs(Expr* expr0, Expr* expr1, bool forward);
 
+  // Can't back prop through merge without making sure one input actually
+  // matches. This can be done on a map or extent basis.
+  // TODO: Move this to val_graph.cpp once validation_utils.cpp is
+  // retired.
+  template <typename T>
+  static bool mapMergeBackward(
```
Comment (zasdfgbnm) [inline csrc/val_graph.h:167]: nit: should this be named as `shouldMapMergeBackward`?

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/id_model/id_model.h (hunk 1 above)
Comment (naoyam) [reply csrc/id_model/id_model.h:59]: Yes, that's what the above "forward through broadcast axes" means.

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Pull request merged**